### PR TITLE
Sort SchemaTransform configuration schema fields by name to establish consistency

### LINKED ISSUE #24361: TypedSchemaTransformProvider returns an inconsistent configuration Schema object
### What happened?

TypedSchemaTransformProvider's `configurationSchema()` method returns Schema object with inconsistent field orderings. A consistent Schema is needed to reliably create configuration Row objects to instantiate and use SchemaTransforms. 

### Issue Priority

Priority: 2

### Issue Component

Component: io-java-gcp

### PR DESCRIPTION
**Pull request by ahmedabu98**
`TypedSchemaTransformProvider::configurationSchema` returns an inconsistently ordered Schema object. A consistent Schema is needed to use configuration Row objects to instantiate and build SchemaTransforms. The solution in this PR is to sort the fields by name.

Fixes #24361

### CODE REVIEW

**Comment by ahmedabu98** [comment]
R: @chamikaramj

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java [hunk 1]
```diff
@@ -61,7 +63,13 @@ Optional<List<String>> dependencies(ConfigT configuration, PipelineOptions optio
   @Override
   public final Schema configurationSchema() {
     try {
-      return SchemaRegistry.createDefault().getSchema(configurationClass());
+      Schema schema = SchemaRegistry.createDefault().getSchema(configurationClass());
+
+      // Sort the fields by name to ensure a consistent schema is produced
+      Schema sortedSchema = schema.getFields().stream()
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java:69]: We should also preserve other fields of the Schema, for example, id, options, encoding_positions_set (and any future fields).

https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/schema.proto#L40

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java [hunk 2]
```diff
@@ -61,7 +63,13 @@ Optional<List<String>> dependencies(ConfigT configuration, PipelineOptions optio
...
+          .sorted(Comparator.comparing(Field::getName))
+          .collect(Schema.toSchema());
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java:71]: Can we also add unit tests  ?

**Review by chamikaramj — COMMENTED** [review]
Thanks.

**Comment by chamikaramj** [comment]
Let's also create a Github issue to track this.

**Comment by ahmedabu98** [comment]
R: @chamikaramj 
Thanks for the feedback, PTAL. I thought the sort code belongs in the Schema class, lmk what you think.

**Comment by ahmedabu98** [comment]
Fixes #24361

**Review by chamikaramj — COMMENTED** [review]
Thanks.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java [hunk 3]
```diff
@@ -40,6 +40,9 @@
  */
 @Internal
 @Experimental(Kind.SCHEMAS)
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20506)
```
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java:44]: No longer running into this now that sorting method is in Schema.java, will remove

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java [hunk 4]
```diff
@@ -264,6 +264,19 @@ public static Schema of(Field... fields) {
     return Schema.builder().addFields(fields).build();
   }
 
+  /** Returns an identical Schema with sorted fields. */
+  public Schema sorted() {
+    Schema sortedSchema =
```
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java:271]: Done, PTAL

**Review by chamikaramj — APPROVED** [review]
LGTM. Thanks.

**Comment by chamikaramj** [comment]
Seems like the new test is failing.

https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/25057/testReport/junit/org.apache.beam.sdk.schemas/SchemaTest/testSortedMethodIncludesAllSchemaFields/

**Comment by ahmedabu98** [comment]
Jacoco creates synthetic fields when it runs, which get picked up by `Class::getDeclaredFields`.

**Comment by ahmedabu98** [comment]
Run Java PreCommit

**Comment by ahmedabu98** [comment]
Run Java PreCommit

**Pull request merged**